store/src/java/com/zimbra/cs/imap/ImapHandler.java

```diff
@@ -2229,41 +2233,84 @@ private boolean doLIST(String tag, String referenceName, Set<String> mailboxName
                 }
             }
 
-            // return only the selected folders (and perhaps their parents) matching the pattern
-            // for owner folders
-            populateFoldersList(ownerPaths, ownerSelected, ownerMatches, returnOptions, remoteSubscriptions, patterns, command, status, selectRecursive, false);
-            // for shared(mounted) folders
-            populateFoldersList(mountPaths, mountSelected, mountMatches, returnOptions, remoteSubscriptions, patterns, command, status, selectRecursive, true);
+            if (ownerSelected.size() <= paginationSize) {
+                ZimbraLog.imap.info("Total folder count is less than folder pagination size.");
+                // return only the selected folders (and perhaps their parents) matching the pattern
+                // for owner folders
+                populateFoldersList(ownerPaths, ownerSelected, ownerMatches, returnOptions, remoteSubscriptions, patterns, command, status, selectRecursive, false);
+                // for shared(mounted) folders
+                populateFoldersList(mountPaths, mountSelected, mountMatches, returnOptions, remoteSubscriptions, patterns, command, status, selectRecursive, true);
+            }
         } catch (ServiceException e) {
             ZimbraLog.imap.warn(command + " failed", e);
             sendNO(tag, command + " failed");
             return canContinue(e);
         }
 
-        // send owners list first
-        if (!ownerMatches.isEmpty()) {
-            for (Object match : ownerMatches.values()) {
-                if (match instanceof String[]) {
-                    for (String response : (String[]) match) {
-                        sendUntagged(response);
-                    }
-                } else {
-                    sendUntagged((String) match);
-                }
-            }
-        }
-
-        // send shared(mounted) folders list
-        if (!mountMatches.isEmpty()) {
-            for (Object match : mountMatches.values()) {
-                if (match instanceof String[]) {
-                    for (String response : (String[]) match) {
-                        sendUntagged(response);
-                    }
-                } else {
-                    sendUntagged((String) match);
-                }
-            }
+        if (ownerSelected.size() > paginationSize) {
```

Comment:
can it be placed as else block for above if?

Reply:
The code sendUntagged was outside the try/catch block, so maintaining it.